docs/style.md

```diff
@@ -196,6 +196,46 @@ external crates. The surprise factor for derivations sharing a name with the sta
 library traits (`Display`) is reduced and it also acts as natural mechanism to tell apart names
 prone to collision (`Serialize`), all without needing to look up the list of imports.
 
+### Visibility
+
+"Minimize visibility" is a standard best practice, but rust has some extra
+specifics. In Rust, there are two granularities for visibility: visibility of
+items within a single crate, and visibility of items across the crates.
+Cross-crate APIs often form important architectural boundaries affecting the
+whole application, while what happens inside a crate concerns only the crate
+itself.
+
+The difference between `pub` and `pub(crate)` is big, while the difference
+between `pub(crate)` and private is relatively small. Be mindful when exposing
+APIs across crates with `pub`. Similarly, be mindful when introducing new
+cross-crate dependencies by editing `Cargo.toml`.
+
+Use `pub` only for items reachable outside of the crate, don't rely on effective
+visibility:
+
+```rust
+// GOOD
+pub(crate) struct Person {
+    pub(crate) first_name: String,
+}
+
+pub(crate) mod auth {
+    pub(crate) check_password(password: &str) -> Result<(), ()>
+}
+
+// BAD
+pub(crate) struct Person {
+    pub first_name: String,
```

Comment:
what is the difference between pub and pub(crate) within a pub(crate) struct?

Reply:
Honestly I don’t like this.  It’s just more syntax and the rationale doesn’t really apply.  I can see how when you’re in a middle of a module you don’t necessarily know whether that module is public or not, but within a struct the start of the struct is just a few lines above.  I’m more sympathetic to `impl` blocks.

Reply:
ok, but what if the intent is: a struct is currently pub(crate), but if it was ever to become pub, all the fields should be pub as well. In that case, having a pub field is a clear indication that whoever the user of the struct is, they should have access to the given field/method. In fact I see hardly any use of pub(crate) fields/methods - either something is an implementation detail of the type, or not.

Reply:
> That isn’t necessarily true just like it’s not necessarily true that all methods of a struct are either private or pub.

That was an assumption of my example, not a conclusion.

Reply:
Got it.  So I guess the complaint here is that the example is poorly labelled.  It’s not bad to use ‘pub’ on fields of a ‘pub(crate)’.  It’s bad to use ‘pub’ on fields which are meant to be crate-public.

Reply:
yes

Reply:
> It’s not bad to use ‘pub’ on fields of a ‘pub(crate)’. It’s bad to use ‘pub’ on fields which are meant to be crate-public.

So this is debatable. It is indeed my intention to declare `pub` field of a `pub(crate)` struct to be bad. And I definitely acknowledge that `pub(crate)` on fields is noisy syntactically. 

My arguments for still using `pub(crate)` even in this scenario are: 

* while it doesn't take a lot of time to look at the struct definition, it takes some attention: you need to combine info from two places
* it breaks "`pub` is always public" invariant, so you go from 0 thinking about pub to epsilon thinking, and to me the difference between 0 and eps seems significant, even if eps is relatively small
* it makes consistency harder: we'd now have to have a rule for *not* using `pub(crate)` on fields unless the struct is `pub`.
* The suggested rule here is enforceable automatically by `-Dunreachable-pub` built-in lint, and that `rustc` lint would complain about pub fields of non-pub struct. I am not entirely sure we should enable this lint (it has some false positives at the moment I think).

Reply:
I had thought about it a little more and while my opinion was at -0 it now shifted to -1 for this PR.  This is because I realised visibility is the least interesting aspect of what `pub` does.  It’s all about level of encapsulation.  When I see `pub` to me it’s signifies that the field, method or function encapsulates internal logic of the type or module and I probably can’t make things inconsistent if I use it.  On the other hand, `pub(crate)` indicates to me that I should avoid using that unless I understand invariants of given type or module.

For example, imagine `struct Complex { pub re: f64, pub im: f64 }`. Making the fields public makes sense because the type will work correctly for whatever finite values someone might put there.  On the other hand, consider `struct Complex { pub(crate) abs: f64, pub(crate) arg: f64 }` where the type has an invariant that 0 ≤ arg < τ.  Now the `pub(crate)` indicates that some abstraction layer is being bypassed if the field is used directly.  This hints that more care needs to be taken when accessing those fields directly.

In yet other words, `pub(crate)` means ‘don’t use unless you’re implementing this type/module’.  If we mandate `pub(crate)` everywhere, we loose that information for crate-private types.

Reply:
Yeah, I think I understand this argument! 

I would say that "encapsulation boundary" and "visibility" are orthogonal. There can be `pub` APIs which are non-the-less escape hatches for breaking abstraction boundaries (ie, `StoreUpdate::set_raw_bytes`). You also don't see visibility at the call-site, so using visibility to mark "this can break invariants" makes the job tough for the reader. It seems that in general all API should maintain invariants, unless the API itself is clearly marked as "this can violate invariant". 

Stated from the user's point of view, if you are using any type's API anywhere outside the implementation of the type, you shouldn't worry about API potentially breaking some invariant: if it could break an invariant, this should be very clear from the API itself.


For the `Complex` example, 

```rust
pub struct Complex { 
  pub(crate) abs: f64, 
  // Invariant:  0 ≤ arg < τ
  pub(crate) arg: f64 
}
```

I would push back against the design that exposes invariant-breaking fields at all. I would strongly prefer the following API instead: 

```rust
pub struct Complex { 
  abs: f64, 
  // Invariant:  0 ≤ arg < τ
  arg: f64 
}

impl Complex {
  pub fn abs(&self) -> f64;
  pub fn with_abs(&self, abs: f64) -> Result<Complex, InvalidAbs>;
  pub fn with_abs_unchecked(&self, abs: f64) -> Complex;

  pub fn arg(&self) -> f64;
  pub fn with_arg(&self, args: f64) -> Result<Complex, InvalidArg>;
  pub fn with_arg_unchecked(&self, args: f64) -> Complex;
}
```

The choice of visibility for `with_arg_unchecked` seems to be orthogonal to the choice of whether to provide this function at all. 

A real world example here would be [`Ratio::new_raw`](https://docs.rs/num-rational/latest/num_rational/struct.Ratio.html#method.new_raw). 

Questions: 

1. Do we need rules here at all? I'd very much like one, as "what should be visibility of a method in a non-pub type" is a frequent question with several plausible answers, and I want to normalize this.
2. If we don't like "pub = inter-crate" for fields&methods, what about the following rule: 
    > Make fields and method either `pub` or private
3. If we want to explicitly allow for invariant-breaking `pub(crate)`, I'd love to see a suggested wording.

Reply:
> The choice of visibility for with_arg_unchecked seems to be orthogonal to the choice of whether to provide this function at all.

This is only true if you exclude the possibility that we want to provide this function to the crate (so that it’s usage is contained and easier to audit) but not as a public API.  I don’t see why ‘everything is either public or private’ would be a good rule to have.

> Do we need rules here at all? I'd very much like one, as "what should be visibility of a method in a non-pub type" is a frequent question with several plausible answers, and I want to normalize this.

We don’t.  The answer to ‘what should be visibility of a method in a non-pub type’ is exactly the same as to ‘what should be visibility of a method in a pub type.’  If it’s part of public interface, make it `pub`.  If it’s not, make it module or crate public.  At this point it becomes question how to organise code (to take advantage of Rust’s visibility rules) rather than about visibility itself.

> Make fields and method either pub or private

This sounds like Java’s philosophy of always providing getters and setters.  IMO it only serves to increase boilerplate code.  Furthermore, since Rust cannot pierce through getters it may make it impossible to take mutable references to multiple fields of a struct.

> If we want to explicitly allow for invariant-breaking pub(crate), I'd love to see a suggested wording.

Invariant-breaking is a bit of a red herring here.  We want to allow `pub(crate)` and whether it’s invariant-breaking or not is separate issue.  And if field, method or function isn’t meant as public interface but code outside of enclosing module needs to access it then it needs to be `pub(crate)`.  There are myriad API combinations so I certainly can’t come up with short guideline encompassing all that.

Reply:
Revised the wording to not have any specific rules for fields and methods, PTAL!